drift.py

```diff
@@ -1,140 +1,77 @@
-            for name, dep in self.last[spec][i].items():
-                cache_map[name] = dep
-        # Iterate through all of the dependencies in the new spec.
-        for i in range(len(concrete_spec)):
-            for name, dep in concrete_spec[i].items():
-                # Check if a dependency has been added.
-                if name not in cache_map:
-                    tags += ["dep-added"]
-                    continue
-                # Check if a dependency version is different.
-                elif dep["full_hash"] != cache_map[name]["full_hash"] and name != spec:
-                    tags += ["dep-updated"]
-                # Remove dependency from known dependencies to track if we don't hit any
-                # dependencies in the new version that use to be in the old spec.
-                cache_map.pop(name)
-        # If there are dependencies that we didn't hit in the new version mark
-        # the commit has having lost dependencies.
-        if len(cache_map) > 0:
-            tags += ["dep-deleted"]
-        return tags
-
-    def tag_variants(self, concrete_spec, spec, name):
-        tags = []
-        if "parameters" in concrete_spec[0][name]:
-            # Check for variants added or modified by comparing new spec to old.
-            for param, value in concrete_spec[0][name]["parameters"].items():
-                if param not in self.last[spec][0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-added"]
-                    else:
-                        tags += ["variant-added"]
-                elif value != self.last[spec][0][name]["parameters"][param]:
-                    if param == "patches":
-                        tags += ["patches-modified"]
-                    else:
-                        tags += ["variant-modified"]
-            # Compare old to new to see if any variants were removed.
-            for param, value in self.last[spec][0][name]["parameters"].items():
-                if param not in concrete_spec[0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-removed"]
-                    else:
-                        tags += ["variant-removed"]
-            return tags
+        for abstract_spec in self.specs:
+            # Separate out spec version and name.
+            verToken = abstract_spec.find("@")
+            if verToken < 0: verToken = len(abstract_spec)
+            out = run(f"spack versions --safe-only {abstract_spec}")
+            # Don't attempt to concretize if the version doesn't yet exist.
+            if abstract_spec[verToken+1:] in out.stdout:
+                out = run(f"spack spec --yaml {abstract_spec}")
+                # If concretization successful check the resulting concrete specs.
+                if out.returncode == SUCCESS:
+                    concrete_spec = spack.spec.Spec().from_yaml(out.stdout)
```

Comment:
Yeah it seems kind of like extra work to be running this twice, once on the command line and now here again?

Reply:
So we actually need two different versions of Spack in the analysis and that's why it's broken down like this. We've got the embedded Spack but we can't also use that version to lookup Specs because we need to change the state of the git repository. So we must use one version of Spack to run the `drift.py` analysis and another within the git repository that we are changing.

Reply:
oh that makes sense. Gotcha.